python/lsst/daf/butler/registries/sqlRegistry.py

```diff
@@ -800,12 +806,81 @@ def addDataUnitEntry(self, dataUnitName, values):
         dataUnit = self._schema.dataUnits[dataUnitName]
         dataUnit.validateId(values)
         dataUnitTable = dataUnit.table
+        v = values.copy()
+        region = v.pop("region", None)
         if dataUnitTable is None:
             raise TypeError("DataUnit '{}' has no table.".format(dataUnitName))
         try:
-            self._connection.execute(dataUnitTable.insert().values(**values))
+            self._connection.execute(dataUnitTable.insert().values(**v))
         except IntegrityError as err:
             raise ValueError(str(err))  # TODO this should do an explicit validity check instead
+        if region is not None:
+            self.setDataUnitRegion((dataUnitName,), v, region, new=True)
+
+    @transactional
+    def setDataUnitRegion(self, dataUnitNames, value, region, new=True):
+        """Set the region field for a DataUnit instance or a combination
+        thereof and update associated spatial join tables.
+
+        Parameters
+        ----------
+        dataUnitNames : sequence
+            A sequence of DataUnit names whose instances are jointly associated
+            with a region on the sky.
+        value : `dict`
+            A dictionary of values that uniquely identify the DataUnits.
+        region : `sphgeom.ConvexPolygon`
+            Region on the sky.
+        new : `bool`
+            If True, the DataUnits associated identified are being inserted for
+            the first time, so no spatial regions should already exist.
+            If False, existing region information for these DataUnits is being
+            replaced.
+        """
+        keyColumns = {}
+        for dataUnitName in dataUnitNames:
+            dataUnit = self._schema.dataUnits[dataUnitName]
+            dataUnit.validateId(value)
+            keyColumns.update(dataUnit.primaryKeyColumns)
+        table = self._schema.dataUnits.getRegionTable(*dataUnitNames)
+        if table is None:
+            raise TypeError("No region table found for '{}'.".format(dataUnitNames))
+        # If a region record for these DataUnits already exists, use an update
+        # query. That could happen either because those DataUnits have been
+        # inserted previously and this is an improved region for them, or
+        # because the region is associated with a single DataUnit instance and
+        # is hence part of that DataUnit's main table.
+        if not new or (len(dataUnitNames) == 1 and table == dataUnit.table):
```

Comment:
I guess `dataUnit` here originates from the loop above? Not illegal but potentially fragile.

Reply:
Right; it looks a bit weird to my C++-trained eyes, too, but I think this is an example of exactly why Python allows loop variables to live beyond their loops.